### PR TITLE
Doc/enterprise#334 - Cluster mode documentation

### PR DESCRIPTION
**Pull request by trgiangdo**
Resolves https://github.com/Avaiga/taipy-enterprise/issues/334

### CODE REVIEW

**Comment by trgiangdo** [comment]
@jrobinAV since the cluster mode feature exists in 3.1, should we also backport this to 3.1 branch?,

That would be better. Indeed.

File: docs/manuals/core/config/job-config.md [hunk 1]
```diff
@@ -114,4 +114,67 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+# Cluster mode
+
+!!! warning "Available in Taipy Enterprise edition"
+
+    This section is relevant only to the Enterprise edition of Taipy.
+
+With the *cluster* mode, a `Job^` runs in its own environment (a worker environment), which can be
+local or on a remote machine, in an asynchronous manner. At the submission, the job is queued in a
+RabbitMQ queue. When there is a worker available, the job is dequeued and sent to the worker
+environment to be executed.
```
Comment (jrobinAV) [inline docs/manuals/core/config/job-config.md]: Trying to add some motivation. What do you think?

```suggestion
The *cluster* mode is designed to make a Taipy application execute the jobs on a remote cluster of distributed workers. The number of machines available in the cluster is unlimited, so the application becomes fully scalable horizontally. It can also be adjusted dynamically depending on some load metric.
With the *cluster* mode, a `Job^` asynchronously runs in its own environment. At the submission, the job is queued in a RabbitMQ queue. When there is a worker available, the job is dequeued and sent to the worker environment to be executed.
```

**Review by jrobinAV — APPROVED** [review]

File: docs/manuals/core/config/job-config.md [hunk 2]
```diff
@@ -114,4 +114,67 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+
+To set up a worker environment, some requirements must be met:
+
+- The worker environment must have access to the same Taipy application as the main environment.
+- The worker environment must have the same Python environment as the main environment.
+- The worker environment must have access to the same RabbitMQ server as the main environment.
+- The worker environment must have access to the data of the data nodes as the main environment.
+  This includes:
+  - The shared drive(s) containing the data for file-based data nodes.
+  - The database server(s) containing the data for database-based data nodes.
```
Comment (jrobinAV) [inline docs/manuals/core/config/job-config.md]: Reformulation. What do you think?

```suggestion
- The main application and the workers must have the same Python environment.
- The main application and the workers must have access to the same RabbitMQ server.
- The main application must be configured to use a Mongo repository, and the workers must have access to it as well.
- The main application and the workers must have access to the data referenced by the the data nodes. This includes:
    - The shared drive(s) containing the data for file-based data nodes.
    - The database server(s) containing the data for database-based data nodes.
```

File: docs/manuals/core/config/job-config.md (hunk 1 above)
Comment (trgiangdo) [reply docs/manuals/core/config/job-config.md]: I realigned the motivation sentence and push a new commit. I think this is a great addition to better explain this mode

File: docs/manuals/core/config/job-config.md (hunk 2 above)
Comment (trgiangdo) [reply docs/manuals/core/config/job-config.md]: The idea behind my proposal is to explain for the user how to set up 1 worker. I call each worker as "worker environment" because the worker can be remote or local, it just needs to be a separate environment.

About the "Mongo repository" requirement, how about "must have access to the entities"?
Mongo is nice to have, but not strictly.

File: docs/manuals/core/config/job-config.md [hunk 3]
```diff
@@ -1,14 +1,14 @@
-The `JobConfig^` allows the developer to configure the Taipy behavior for job executions. Two main modes are
-available in Taipy: the `standalone` mode and the `development` mode.
+The `JobConfig^` allows the developer to configure the Taipy behavior for job executions. There are three modes
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
`JobConfig^` allows the developer to configure the Taipy behavior for job executions. There are three modes
```
or
```suggestion
The `JobConfig^` class allows the developer to configure the Taipy behavior for job executions. There are three modes
```

File: docs/manuals/core/config/job-config.md [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
+available in Taipy: the *standalone* mode, the *development* mode, and the *cluster* mode
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
available in Taipy: *standalone*, *development*, and *cluster*.
```

File: docs/manuals/core/config/job-config.md [hunk 5]
```diff
@@ -54,10 +54,10 @@ The *development* mode can be activated with the following config:
 With the *standalone* mode, a `Job^` runs in its own execution context, in an asynchronous manner.
 At the submission, the job is queued. It is dequeued in a different thread and sent to a dedicated process to be
 executed.
-Note that with the *standalone* mode, the submit method in not blocking and returns after the job is queued.
+Note that with the *standalone* mode, the submit method is not blocking and returns after the job is queued.
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
Note that with the *standalone* mode, the `(Scenario.)submit()^` method is not blocking and returns after the job is queued.
```

File: docs/manuals/core/config/job-config.md [hunk 6]
```diff
@@ -114,4 +114,71 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
     max_nb_of_workers = "8:int"
     ```
 
+# Cluster mode
+
+!!! warning "Available in Taipy Enterprise edition"
+
+    This section is relevant only to the Enterprise edition of Taipy.
+
+The *cluster* mode is designed to make a Taipy application executes jobs on a remote cluster of
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
The *cluster* mode is designed to make Taipy applications execute jobs on a remote cluster of
```

File: docs/manuals/core/config/job-config.md [hunk 7]
```diff
@@ -114,4 +114,71 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+distributed workers. The number of machines available in the cluster is unlimited, so the
+application becomes fully scalable horizontally. It can also be adjusted dynamically depending on
+some load metric.
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
some load metrics.
```

File: docs/manuals/core/config/job-config.md [hunk 8]
```diff
@@ -114,4 +114,71 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+
+With the *cluster* mode, a `Job^` asynchronously runs in its own environment. At the submission,
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
With the *cluster* mode, a `Job^` asynchronously runs in its own environment. When submitted,
```

File: docs/manuals/core/config/job-config.md [hunk 9]
```diff
@@ -114,4 +114,71 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+the job is queued in a RabbitMQ queue. When there is a worker available, the job is dequeued and
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md:129]: Do we need to mention the technology (RabbitMQ) at this point?

File: docs/manuals/core/config/job-config.md [hunk 10]
```diff
@@ -114,4 +114,71 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+sent to the worker environment to be executed.
+
+Note that with the *cluster* mode, the submit method is not blocking and returns after the job is queued.
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
Note that with the *cluster* mode, the `(Scenario.)submit()^` method is not blocking and returns after the job is queued.
```

File: docs/manuals/core/config/job-config.md [hunk 11]
```diff
@@ -114,4 +114,71 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+It means the submit method can return before the job finishes or even before it is dequeued. Please refer to
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]:
```suggestion
This means the `submit()` method can return before the job finishes or even before it is dequeued. Please refer to
```

File: docs/manuals/core/config/job-config.md [hunk 12]
```diff
@@ -114,4 +114,71 @@ For example, the following configuration will allow Taipy to run up to eight `Jo
...
+    ```toml linenums="1" title="config.toml"
+    [JOB]
+    mode = "cluster"
+    ```
+
+To set up a worker environment, some requirements must be met:
+
+- The worker environment must have access to the same Taipy application as the main environment.
+- The worker environment must have the same Python environment as the main environment.
+- The worker environment must have access to the same RabbitMQ server as the main environment.
```
Comment (FabienLelaquais) [inline docs/manuals/core/config/job-config.md]: Do we need to expose the techno here?

**Review by FabienLelaquais — APPROVED** [review]
Minor improvements proposed...

File: docs/manuals/core/config/job-config.md (hunk 9 above)
Comment (trgiangdo) [reply docs/manuals/core/config/job-config.md:129]: We probably not. What do you think @jrobinAV?

File: docs/manuals/core/config/job-config.md (hunk 12 above)
Comment (trgiangdo) [reply docs/manuals/core/config/job-config.md]: Most is necessary I think. However, for the "RabbitMQ", it may be too specific. What do you think @jrobinAV?

File: docs/manuals/core/config/job-config.md (hunk 9 above)
Comment (jrobinAV) [reply docs/manuals/core/config/job-config.md:129]: I would say yes. 
We rely on RabbitMQ. Users need to know it to start it and manage it.

File: docs/manuals/core/config/job-config.md (hunk 12 above)
Comment (jrobinAV) [reply docs/manuals/core/config/job-config.md]: That is a prerequisite. We rely on RabbitMQ for this feature. I would keep it here.

File: docs/manuals/core/config/job-config.md (hunk 2 above)
Comment (jrobinAV) [reply docs/manuals/core/config/job-config.md]: 1. **About worker environment:** For me, the whole feature makes sense only on remote machines. It does not make sense to me to run a local worker. We have the standalone mode for that. This is only useful for us as contributors, but not really for our users. What do you think?

2. **About mongo:** I thought a `mongo` repository was mandatory to run in `cluster` mode. Is it possible to run a remote worker on a filesystem repository, for instance? If yes, I agree with your proposal.

Comment (trgiangdo) [reply docs/manuals/core/config/job-config.md]: 1. That makes sense. Let me rephrase this
2. Yes, the user can use cluster mode as long as all machines have access to the same entity repository. For FS and SQLite, it can be a shared folder or file. For Mongo, it's more convinient
